### PR TITLE
Mon 4046 poller attached to multiple remote servers

### PR DESCRIPTION
**Pull request by lpinsivy**
#  Description

This feature allows by using poller configuration wizard or be editing an existing poller to attache this one to additional Remote Servers.

By using the wizard and by selecting additional Remote Servers, the broker configuration will be configured to send collected data to the Master Remote Server and to the additional Remote Servers.

More over, the Centcore process can use additional Remote Servers to send configuration or to forward external commands if the Master Remote Server is not available during the process.

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [x] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [ ] 19.04.x
- [x] 19.10.x (master)

<h2> How this pull request can be tested ? </h2>

- Create a simple poller using configuration wizard: the associated broker output must be defined to connect to the Centreon Central Server
- Create a simple poller using configuration wizard and attach it to a Master Remote Server: the associated broker output must be defined to connect to the Remote Server
- Create a simple poller using configuration wizard and attach it to a Master Remote Server and additional Remote Servers (the associated broker outputs must be defined to connect to the Master Remote Server) and the additional Remote Servers.
- Need to test to send external commands and configuration by selecting "No" for "Use the Remote Server as a proxy for SSH" in poller form.
- Need to generate / export Centreontrapd database (with SNMP trap definition linked to a service of a resource monitored by the poller) with or without "Use the Remote Server as a proxy for SSH" in poller form.

#### Community contributors & Centreon team

- [x] I followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

#### Centreon team only

- [x] I have made sure that the **unit tests** related to the story are successful.
- [x] I have made sure that **unit tests cover 80%** of the code written for the story.
- [x] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

### CODE REVIEW

File: www/install/createTables.sql [hunk 1]
```diff
@@ -2296,6 +2296,12 @@ CREATE TABLE IF NOT EXISTS `remote_servers` (
   `no_proxy` enum('0','1') NOT NULL DEFAULT '0'
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 
+-- Create rs_poller_relation for the additional relationship between poller and remote servers
+CREATE TABLE `rs_poller_relation` (
+  `remote_server_id` int(11) NOT NULL DEFAULT '0',
```
Comment (kduret) [inline www/install/createTables.sql]:
```suggestion
  `remote_server_id` int(11) NOT NULL DEFAULT 0,
```

File: www/install/createTables.sql [hunk 2]
```diff
@@ -2296,6 +2296,12 @@ CREATE TABLE IF NOT EXISTS `remote_servers` (
...
+  `poller_server_id` int(11) NOT NULL DEFAULT '0',
```
Comment (kduret) [inline www/install/createTables.sql]:
```suggestion
  `poller_server_id` int(11) NOT NULL DEFAULT 0,
```

File: www/install/sql/centreon/Update-DB-19.10.0-beta.3.sql [hunk 3]
```diff
@@ -6,3 +6,10 @@ ALTER TABLE nagios_server ADD COLUMN `remote_server_centcore_ssh_proxy` enum('0'
 -- Add severity preference on host-monitoring widget
 INSERT INTO `widget_parameters_field_type` (`ft_typename`, `is_connector`) VALUES
 ('hostSeverityMulti', 1);
+
+-- Create rs_poller_relation for the additional relationship between poller and remote servers
+CREATE TABLE `rs_poller_relation` (
+  `remote_server_id` int(11) NOT NULL DEFAULT '0',
```
Comment (kduret) [inline www/install/sql/centreon/Update-DB-19.10.0-beta.3.sql]:
```suggestion
  `remote_server_id` int(11) NOT NULL DEFAULT 0,
```

File: www/install/sql/centreon/Update-DB-19.10.0-beta.3.sql [hunk 4]
```diff
@@ -6,3 +6,10 @@ ALTER TABLE nagios_server ADD COLUMN `remote_server_centcore_ssh_proxy` enum('0'
...
+  `poller_server_id` int(11) NOT NULL DEFAULT '0',
```
Comment (kduret) [inline www/install/sql/centreon/Update-DB-19.10.0-beta.3.sql]:
```suggestion
  `poller_server_id` int(11) NOT NULL DEFAULT 0,
```

File: www/install/createTables.sql (hunk 1 above)
Comment (kduret) [reply www/install/createTables.sql]: In fact, I think it should not have default value

File: www/install/createTables.sql (hunk 2 above)
Comment (kduret) [reply www/install/createTables.sql]: In fact, I think it should not have default value

File: www/install/sql/centreon/Update-DB-19.10.0-beta.3.sql (hunk 3 above)
Comment (kduret) [reply www/install/sql/centreon/Update-DB-19.10.0-beta.3.sql]: In fact, I think it should not have default value

File: www/install/sql/centreon/Update-DB-19.10.0-beta.3.sql (hunk 4 above)
Comment (kduret) [reply www/install/sql/centreon/Update-DB-19.10.0-beta.3.sql]: In fact, I think it should not have default value

File: www/include/configuration/configServers/DB-Func.php [hunk 5]
```diff
@@ -328,6 +373,33 @@ function duplicateServer(array $server, array $nbrDup): void
...
+    global $pearDB;
+
+    $dbResult = $pearDB->query(
+        "DELETE FROM rs_poller_relation WHERE poller_server_id = '" . $id . "'"
+    );
+
+    if (!is_null($remotes)) {
+        foreach ($remotes as $remote) {
+            $dbResult = $pearDB->query(
+                "INSERT INTO rs_poller_relation VALUES ('" . $remote . "', '" . $id  . "')"
```
Comment (sc979) [inline www/include/configuration/configServers/DB-Func.php]: please could you bind these values

File: www/include/configuration/configServers/formServers.php [hunk 6]
```diff
@@ -74,11 +75,33 @@
     }
 
     if ($serverType === "remote") {
-        $dbResult = $pearDB->query("SELECT http_method, http_port, no_check_certificate, no_proxy " .
-            "FROM `remote_servers` WHERE `ip` = '" . $cfg_server['ns_ip_address'] . "' LIMIT 1");
+        $dbResult = $pearDB->query(
```
Comment (sc979) [inline www/include/configuration/configServers/formServers.php]: please could use a prepared statement and bind $cfg_server value

File: www/include/configuration/configServers/formServers.php [hunk 7]
```diff
@@ -74,11 +75,33 @@
...
+            FROM `remote_servers` 
+            WHERE `ip` = '" . $cfg_server['ns_ip_address'] . "' LIMIT 1"
+        );
         $cfg_server = array_merge($cfg_server, array_map("myDecode", $dbResult->fetch()));
         $dbResult->closeCursor();
     }
+
+    if ($serverType === "poller") {
+        // Select additionnal Remote Servers
+        $dbResult = $pearDB->query(
```
Comment (sc979) [inline www/include/configuration/configServers/formServers.php]: please could use a prepared statement and bind $cfg_server value

File: www/include/configuration/configServers/formServers.php [hunk 8]
```diff
@@ -74,11 +75,33 @@
...
+            "SELECT remote_server_id, name 
+            FROM rs_poller_relation AS rspr
+            LEFT JOIN nagios_server AS ns ON (rspr.remote_server_id = ns.id)
+            WHERE poller_server_id = '" . $cfg_server['id'] . "'"
+        );
+        if ($dbResult->numRows() > 0) {
+            while ($row = $dbResult->fetchRow()) {
```
Comment (sc979) [inline www/include/configuration/configServers/formServers.php]:
```suggestion
            while ($row = $dbResult->fetch()) {
```

File: www/include/configuration/configServers/formServers.php [hunk 9]
```diff
@@ -293,8 +329,14 @@
  * Form Rules
  */
 $form->registerRule('exist', 'callback', 'testExistence');
+$form->registerRule('testAdditionalRemoteServer', 'callback', 'testAdditionalRemoteServer');
 $form->addRule('name', _("Name is already in use"), 'exist');
 $form->addRule('name', _("The name of the poller is mandatory"), 'required');
+$form->addRule(
+    array('remote_additional_id', 'remote_id'),
+    _('To use additional Remote Servers a Master Remote Server must be selected.'),
```
Comment (sc979) [inline www/include/configuration/configServers/formServers.php]: please add translated sentences / template

File: www/include/configuration/configServers/help.php [hunk 10]
```diff
@@ -4,7 +4,8 @@
 $help['ns_ip_address'] = dgettext("help", "IP address of the poller");
 $help['localhost'] = dgettext("help", "Whether the poller is local");
 $help['is_default'] = dgettext("help", "Main poller");
-$help['remote_id'] = dgettext("help", "Remote Server to which this server will be attached");
+$help['remote_id'] = dgettext("help", "Master Remote Server to which this server will be attached");
```
Comment (sc979) [inline www/include/configuration/configServers/help.php]: please add the translated sentences or template

File: www/install/react_translation.php [hunk 11]
```diff
@@ -1,6 +1,6 @@
 <?php
 /*
- * Copyright 2005-2018 Centreon
+ * Copyright 2005-2019 Centreon
  * Centreon is developped by : Julien Mathis and Romain Le Merlus under
```
Comment (sc979) [inline www/install/react_translation.php]:
```suggestion
 * Centreon is developed by : Julien Mathis and Romain Le Merlus under
```

File: www/include/configuration/configServers/formServers.php (hunk 9 above)
Comment (lpinsivy) [reply www/include/configuration/configServers/formServers.php]: the messages.pot file should be generated from all sources automatically

File: www/include/configuration/configServers/help.php (hunk 10 above)
Comment (lpinsivy) [reply www/include/configuration/configServers/help.php]: the messages.pot file should be generated from all sources automatically

File: www/include/configuration/configServers/formServers.php (hunk 9 above)
Comment (sc979) [reply www/include/configuration/configServers/formServers.php]: I meant -> add the new sentence to the translation files. And please translate it in the french one ;)

**Pull request merged**